loclist.kak

```diff
@@ -0,0 +1,147 @@
+    *diagnostics* *goto* *grep* *implementations* *lint-output* *make* *references* *symbols*
+
+    %opt{loclist_buffer} determines the buffer current location list buffer.
+    %opt{loclist_location_format} determines matching locations.
+} -buffer-completion %{
+    evaluate-commands -try-client %opt{jumpclient} -save-regs / %{
+        buffer %opt{loclist_buffer}
+        set-register / %opt{loclist_location_format}
+        execute-keys ge %opt{grep_current_line}g<a-h> <a-/><ret>
+        loclist-jump
+    }
+    try %{
+        evaluate-commands -client %opt{toolsclient} %{
+            buffer %opt{loclist_buffer}
+            execute-keys gg %opt{grep_current_line}g
+        }
+    }
+}
+
+define-command -override loclist-stack-push -docstring "record location list buffer" %{
+    evaluate-commands %sh{
+        eval set -- $kak_quoted_opt_loclist_stack
+        if printf '%s\n' "$@" | grep -Fxq -- "$kak_bufname"; then {
+            exit
+        } fi
+        newbuf=$kak_bufname-$#
+        echo "try %{ delete-buffer! $newbuf }"
+        echo "rename-buffer $newbuf"
+        echo "set-option -add global loclist_stack %val{bufname}"
+    }
+    set-option global loclist_buffer %val{bufname}
+}
+
+define-command -override loclist-stack-pop -docstring "restore location list buffer" %{
+    evaluate-commands %sh{
+        eval set -- $kak_quoted_opt_loclist_stack
+        if [ $# -eq 0 ]; then {
+            echo fail "loclist-stack-pop: no location list buffer to pop"
+            exit
+        } fi
+        printf 'set-option global loclist_stack'
+        top=
+        while [ $# -ge 2 ]; do {
+            top=$1
+            printf ' %s' "$1"
+            shift
+        } done
+        echo
+        echo "delete-buffer $1"
+        echo "set-option global loclist_buffer '$top'"
+    }
+    try %{
+        evaluate-commands -try-client %opt{jumpclient} %{
+            buffer %opt{loclist_buffer}
+            grep-jump
+        }
+    }
+}
+define-command -override loclist-stack-clear -docstring "clear location list buffers" %{
+    evaluate-commands %sh{
+        eval set --  $kak_quoted_opt_loclist_stack
+        printf 'try %%{ delete-buffer %s }\n' "$@"
+    }
+    set-option global loclist_stack
+}
+
+# TODO: What exactly does this disable, and do we need to?
+set-option global disabled_hooks "%opt{disabled_hooks}|grep-jump"
+declare-option regex loclist_buffer_regex \*(?:callees|callers|diagnostics|goto|find|grep|implementations|lint-output|references)\*(?:-\d+)?
+
+hook -group loclist global WinDisplay %opt{loclist_buffer_regex} %{
+    loclist-stack-push
+}
+
+hook -group loclist global GlobalSetOption '^loclist_buffer=(.*)$' %{
+    # TODO: Upstream as loclist-write?
+    # alias "buffer=%opt{loclist_buffer}" w grep-write
+    map "buffer=%opt{loclist_buffer}" r
+}
+
+# NOTE: Only relevant if other bufer-stack systems are included
+# alias global buffers-pop loclist-stack-pop
+# alias global buffer-clear loclist-stack-clear
```

Comment:
I mapped `git-stack-pop` to `git q` (Magit uses `q`).
Mainly because it was easy.

I think from a UI standpoint, a context-sensitive back-button makes a ton of sense.

My proposal is that the global aliases point to `loclist-*` while in `*git*` buffers we shadow that with the `git-stack-*` commands.

It would be even better if we found a good key mapping for the pop-command but I don't think it's essential.

Reply:
I Agree - we may as well introduce these concepts here